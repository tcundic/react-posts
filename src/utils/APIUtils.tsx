import {apiEndpoint} from "../config";
import RequestOptions from "../interfaces/Request";
import Post from "../interfaces/Post";

const request = (options: RequestOptions, additionalHeaders?: Object): Promise<any> => {
    const headers = {
        'Content-Type': 'application/json',
        ...additionalHeaders
    };

    const defaults = {headers};
    options = Object.assign({}, defaults, options, {crossDomain: true});

    return fetch(options.url, options)
        .then(response => response.json());
};

export function getPosts(): Promise<Post[]> {
    return request({
        url: `${apiEndpoint}posts`,
        method: 'GET'
    });
}

export function getPostById(id: number): Promise<Post> {
    return request({
        url: `${apiEndpoint}posts/${id}`,
        method: 'GET'
    });
}
