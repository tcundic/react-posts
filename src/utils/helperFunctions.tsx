export const greet = (greeting: string, componentName: string): void => console.log(`${greeting} ${componentName}`);