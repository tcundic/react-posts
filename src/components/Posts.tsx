import {useEffect, useState} from "react";
import {greet} from "../utils/helperFunctions";
import {getPosts} from "../utils/APIUtils";
import Post from "../interfaces/Post";
import PostList from "./PostList";
import useFilter from "./useFilter";

const Posts = ({greeting}: {greeting: string}) => {

    const [posts, setPosts] = useState<Post[]>([]);
    const [filteredPosts, setFilteredPosts] = useState<Post[]>(posts);
    const [username, usernameInput] = useFilter({greeting, type: "text", placeholder: "Enter username"});

    useEffect(() => {
        greet(greeting, Posts.name);
    }, [greeting]);

    useEffect(() => {
        getPosts().then((posts: Post[]) => {
            setPosts(posts);
            setFilteredPosts(posts);
        });
    }, []);

    useEffect(() => {
        // @ts-ignore
        setFilteredPosts(posts.filter((post: Post) => post.user.includes(username)));
    }, [username, posts]);

    return (
        <div className="mt-6 mx-6">
            <div className="posts-header">
                <h1>Posts</h1>
                {usernameInput}
            </div>
            <PostList posts={filteredPosts} greeting={greeting} />
        </div>
    );
};

export default Posts;