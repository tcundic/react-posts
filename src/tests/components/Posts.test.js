import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as Greet from "../../utils/helperFunctions";
import React from "react";
import Posts from "../../components/Posts";
import posts from "../fixtures/posts";
import { BrowserRouter } from "react-router-dom";
import { act } from "@testing-library/react";

beforeEach(() => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
        json: jest.fn().mockResolvedValue(posts)
    })
});

afterEach(() => {
    jest.restoreAllMocks();
});

const waitForComponentToPaint = async (wrapper) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
        wrapper.update();
    });
};

const greeting = "Hello from";
const name = Posts.name;

test('should render Posts correctly', () => {
    const wrapper = shallow(<Posts greeting={greeting} />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

test('should print Posts greeting', () => {
    const greetSpy = jest.spyOn(Greet, 'greet');
    const wrapper = mount(<BrowserRouter><Posts greeting={greeting}/></BrowserRouter>);
    waitForComponentToPaint(wrapper);
    expect(greetSpy).toHaveBeenLastCalledWith(greeting, name);
});
