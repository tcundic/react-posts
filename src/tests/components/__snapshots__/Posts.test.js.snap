// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Posts correctly 1`] = `
<div
  className="mt-6 mx-6"
>
  <div
    className="posts-header"
  >
    <h1>
      Posts
    </h1>
    <input
      onChange={[Function]}
      placeholder="Enter username"
      type="text"
      value=""
    />
  </div>
  <PostList
    greeting="Hello from"
    posts={Array []}
  />
</div>
`;
